ndscan/experiment/fragment.py

```diff
@@ -641,13 +705,107 @@ def get_default_analyses(self) -> Iterable[DefaultAnalysis]:
+    device as well to avoid costly kernel recompilations at each point.
+    """
+    def build_fragment(self, exp_fragments: List[ExpFragment]) -> None:
+        """
+        :param exp_fragments: The "child" fragments to execute. The fragments will be
+            run in the given order. No special treatment is given to the
+            ``{host,device}_{setup,cleanup}()`` methods, which by default will just be
+            executed using the default recursive behaviour.
+        """
+        if not exp_fragments:
+            raise ValueError("At least one child ExpFragment should be given")
+        self._exp_fragments = exp_fragments
+
+        is_kernels = [is_kernel(e.run_once) for e in exp_fragments]
+        if all(is_kernels):
+            self.run_once = kernel(self.run_once)
+        else:
+            if any(is_kernels):
+                logger.warning("Mixed host/@kernel run_once() methods among passed " +
+                               "ExpFragments; execution will be slow as the " +
+                               "kernel(s) will be recompiled for each scan point.")
+
+        # Since polymorphism is not supported by the ARTIQ compiler, make named
+        # attributes for the exp fragments and implement run_once() by explicitly
+        # calling them one after each other.
+        for i, exp in enumerate(exp_fragments):
+            setattr(self, f"_exp_fragment_{i}", exp)
+        self._run_once_impl = kernel_from_string(["self"], "\n".join(
+            [f"self._exp_fragment_{i}.run_once()" for i in range(len(exp_fragments))]),
+                                                 portable)
+
+    def prepare(self) -> None:
+        ""
+        for exp in self._exp_fragments:
+            exp.prepare()
+
+    def run_once(self) -> None:
+        return self._run_once_impl(self)
+
+    def get_always_shown_params(self) -> List[ParamHandle]:
+        """Collect always-shown params from each child fragment, plus any parameters
+        directly defined in this fragment as usual.
+        """
+        result = super().get_always_shown_params()
+        for exp in self._exp_fragments:
+            result += exp.get_always_shown_params()
+        return result
+
+    def get_default_analyses(self) -> Iterable[DefaultAnalysis]:
+        """Collect default analyses from each child fragment.
+
+        The analyses are wrapped in a proxy that prepends any result channel names with
+        the fragment path to ensure results from different analyses do not collide.
+        """
+        analyses = []
+        for exp in self._exp_fragments:
+            exp_analyses = exp.get_default_analyses()
+            prefix = "_".join(
+                _skip_common_prefix(exp._fragment_path, self._fragment_path)) + "_"
+            analyses += [
+                ResultPrefixAnalysisWrapper(analysis, prefix)
+                for analysis in exp_analyses
+            ]
+        return analyses
```

Comment:
`OnlineAnalysis` are displayed correctly for all child `ExpFragment`s, but `CustomAnalysis` are not displayed.

Reply:
Couldn't reproduce with a simple test case:

```py
from ndscan.experiment import *
import numpy as np


class Foo(ExpFragment):

    def build_fragment(self) -> None:
        self.setattr_param("x", FloatParam, "X", 0.0)
        self.setattr_result("y", FloatChannel)

    def run_once(self) -> None:
        self.y.push(self.x.get())

    def get_default_analyses(self):
        return [
            CustomAnalysis([self.x], lambda params, results, channels: [
                Annotation(
                    "location",
                    coordinates={self.x: np.mean(params[self.x])},
                )
            ])
        ]


ScanFoo = make_fragment_scan_exp(Foo)


class TwoFoos(AggregateExpFragment):

    def build_fragment(self):
        self.setattr_fragment("a", Foo)
        self.setattr_fragment("b", Foo)
        self.setattr_param_rebind("val", self.a, "x")
        self.b.bind_param("x", self.val)
        super().build_fragment([self.a, self.b])


ScanTwoFoos = make_fragment_scan_exp(TwoFoos)
```

How exactly does your setup look?